### ISSUE TITLE
Building on a Cray system with OpenMP enabled fails

### ISSUE THREAD

**Issue by finkandreas** [kokkos/kokkos#2785]
I'm trying to build Kokkos on a cray system with their variant of Clang, and it fails to detect OpenMP.
`cmake -DKokkos_ENABLE_OPENMP src_dir` fails with the error message:
```
CMake Error at cmake/kokkos_corner_cases.cmake:20 (MESSAGE):
  Clang failed OpenMP check.  You have requested -DKokkos_ENABLE_OPENMP=ON,
  but the Clang compiler does not appear to have been built with OpenMP
  support
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:178 (INCLUDE)
  CMakeLists.txt:146 (KOKKOS_SETUP_BUILD_ENVIRONMENT)
```

Looking at the file `cmake/kokkos_corner_cases.cmake` it seems that you are assuming that clang is using libomp as OpenMP implementation, however this is not true for the cray version. When I try to compile a simple test program with OpenMP with `CC -fopenmp test.cpp` and look at the output of ldd I see the following:
```
        linux-vdso.so.1 (0x00007fff6cbe0000)
        /apps/daint/UES/xalt/xalt2/software/xalt/2.7.24/lib64/libxalt_init.so (0x00002b69bedf7000)
        libAtpSigHandler.so.0 => /opt/cray/pe/lib64/libAtpSigHandler.so.0 (0x00002b69bf00c000)
        librca.so.0 => /opt/cray/rca/default/lib64/librca.so.0 (0x00002b69bf216000)
        libquadmath.so.0 => /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libquadmath.so.0 (0x00002b69bf41a000)
        libmodules.so.1 => /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libmodules.so.1 (0x00002b69bf65a000)
        libfi.so.1 => /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libfi.so.1 (0x00002b69bf875000)
        libcraymath.so.1 => /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libcraymath.so.1 (0x00002b69bff97000)
        libf.so.1 => /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libf.so.1 (0x00002b69c0270000)
        libu.so.1 => /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libu.so.1 (0x00002b69c0503000)
        libcsup.so.1 => /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libcsup.so.1 (0x00002b69c0807000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00002b69c0a0d000)
        libm.so.6 => /lib64/libm.so.6 (0x00002b69c0c2b000)
        libstdc++.so.6 => /opt/cray/pe/gcc-libs/libstdc++.so.6 (0x00002b69c0f63000)
        libcraymp.so.1 => /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libcraymp.so.1 (0x00002b69c12ec000)
        libgcc_s.so.1 => /opt/cray/pe/gcc-libs/libgcc_s.so.1 (0x00002b69c15bb000)
        libc.so.6 => /lib64/libc.so.6 (0x00002b69c17d3000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00002b69c1b8d000)
        librt.so.1 => /lib64/librt.so.1 (0x00002b69c1d91000)
        libgfortran.so.5 => /opt/cray/pe/gcc-libs/libgfortran.so.5 (0x00002b69c1f99000)
        /lib64/ld-linux-x86-64.so.2 (0x00002b69bebcf000)
        libz.so.1 => /lib64/libz.so.1 (0x00002b69c2408000)
```

Looking inside the library `libcraymp.so.1` reveals that it contains the OpenMP symbols:
```
anfink@daint102 ~/tmp/kokkos.20200221/build $ objdump -T /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/libcraymp.so.1 | grep omp_get_num_threads
0000000000038150 g    DF .text  00000000000000ee  OMP_1.0     omp_get_num_threads_
0000000000038150 g    DF .text  00000000000000ee  OMP_1.0     omp_get_num_threads
```

**Comment by jjwilke** [comment]
When you say the Cray variant of Clang - do you mean the Cray compiler (9.0) that uses a Clang frontend or the default Clang compiler on a Cray system.

And pending that, can you post the whole output of the configure? Ideally the Cray compiler would identify as Cray, not Clang.

-CMake version?
-Platform?

**Comment by finkandreas** [comment]
I mean the cray compiler that uses a Clang frontend, sorry for not being precise here.
`CMAKE_CXX_COMPILER` is set to `CC`
```
anfink@daint102 ~ $ CC --version
Cray clang version 9.0.2 (2a3a7003aaa5b93e2070bde59a5ee6b9682b67d7) (based on LLVM 9.0.0svn)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/cray/pe/cce/9.0.2/cce-clang/x86_64/bin
```
or as an alternative:
```
anfink@daint102 ~ $ CC -v
Cray clang version 9.0.2 (2a3a7003aaa5b93e2070bde59a5ee6b9682b67d7) (based on LLVM 9.0.0svn)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/cray/pe/cce/9.0.2/cce-clang/x86_64/bin
Found candidate GCC installation: /opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0
Selected GCC installation: /opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
 "/opt/cray/pe/cce/9.0.2/binutils/x86_64/x86_64-pc-linux-gnu/bin/ld.gold" --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o a.out /usr/lib/../lib64/crt1.o /usr/lib/../lib64/crti.o /opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0/crtbegin.o -L/opt/cray/dmapp/default/lib64 -L/opt/cray/pe/mpt/7.7.10/gni/mpich-crayclang/9.0/lib -L/opt/cray/dmapp/default/lib64 -L/opt/cray/pe/mpt/7.7.10/gni/mpich-crayclang/9.0/lib -L/opt/cray/pe/libsci/19.06.1/CRAYCLANG/9.0/x86_64/lib -L/opt/cray/pe/netcdf-hdf5parallel/4.6.3.1/crayclang/9.0/lib -L/opt/cray/pe/hdf5-parallel/1.10.5.1/crayclang/9.0/lib -L/opt/cray/rca/2.2.20-7.0.1.1_4.9__g8e3fb5b.ari/lib64 -L/opt/cray/pe/atp/2.1.3/libApp -L/opt/cray/pe/cce/9.0.2/cce/x86_64/lib/pkgconfig/../ -L/opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0 -L/opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0/../../.. -L/opt/cray/pe/cce/9.0.2/cce-clang/x86_64/bin/../lib -L/lib -L/usr/lib -rpath=/opt/cray/pe/cce/9.0.2/cce/x86_64/lib -rpath=/opt/cray/pe/gcc-libs --no-as-needed -lAtpSigHandler -lAtpSigHCommData --undefined=_ATP_Data_Globals --undefined=__atpHandlerInstall -lrca --as-needed -lnetcdf_c++4 --no-as-needed --as-needed -lnetcdf --no-as-needed --as-needed -lhdf5_hl_parallel --no-as-needed --as-needed -lhdf5_parallel --no-as-needed --as-needed -lmpich_crayclang --no-as-needed --as-needed -lmpichcxx_crayclang --no-as-needed --as-needed -lsci_cray_mpi --no-as-needed --as-needed -lsci_cray --no-as-needed --as-needed -lpgas-dmapp --no-as-needed -lquadmath -lmodules -lfi -lcraymath -lf -lu -lcsup --as-needed -lpthread -latomic --no-as-needed --as-needed -lm --no-as-needed /opt/cray/pe/cce/9.0.2/cce/x86_64/lib/no_mmap.o -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /opt/gcc/8.1.0/snos/lib/gcc/x86_64-suse-linux/8.1.0/crtend.o /usr/lib/../lib64/crtn.o
```

CMake:
```
anfink@daint102 ~ $ cmake --version
cmake version 3.14.5
```

Platform:
```
anfink@daint102 ~ $ cat /etc/os-release 
NAME="SLES"
VERSION="15"
VERSION_ID="15"
PRETTY_NAME="SUSE Linux Enterprise Server 15"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15"
```

Full configure output:
```
anfink@daint102 ~/tmp/kokkos.20200221/build $ cmake -DCMAKE_CXX_COMPILER=CC -DKokkos_ENABLE_OPENMP=ON ../
-- Setting default Kokkos CXX standard to 11
-- The CXX compiler identification is Clang 9.0.0
-- Check for working CXX compiler: /opt/cray/pe/craype/2.6.1/bin/CC
-- Check for working CXX compiler: /opt/cray/pe/craype/2.6.1/bin/CC -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The project name is: Kokkos
-- Using -std=gnu++11 for C++11 extensions as feature
-- Execution Spaces:
--   Device Parallel: NONE
--     Host Parallel: OPENMP
--       Host Serial: NONE
-- 
-- Architectures:
-- Found TPLLIBDL: /usr/lib64/libdl.so  
CMake Error at cmake/kokkos_corner_cases.cmake:20 (MESSAGE):
  Clang failed OpenMP check.  You have requested -DKokkos_ENABLE_OPENMP=ON,
  but the Clang compiler does not appear to have been built with OpenMP
  support
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:178 (INCLUDE)
  CMakeLists.txt:146 (KOKKOS_SETUP_BUILD_ENVIRONMENT)


-- Configuring incomplete, errors occurred!
See also "/users/anfink/tmp/kokkos.20200221/build/CMakeFiles/CMakeOutput.log".
```

**Comment by jjwilke** [comment]
Huh. Okay. We will have to think about how to fix that. I didn't realize the newer Cray compilers would identify to CMake as Clang.

At the very least, I can put in a hook to make OpenMP flags override-able so people can work around it.

**Comment by crtrott** [comment]
I am looking into this right now. As reproducer on voltrino:
```
crtrott@voltrino:~> module swap PrgEnv-intel/6.0.5 PrgEnv-cray/6.0.6 
crtrott@voltrino:~> module swap cce/8.7.9 cce/9.1.2
crtrott@voltrino:~> module unload cray-libsci/19.02.1 
crtrott@voltrino:~> module unload cray-mpich/7.7.6 
```

**Comment by PhilMiller** [comment]
Likely resolved by #4105, since that reports explicit testing of the use case here.
